File: android/app/src/main/java/org/findmyfam/models/InviteCode.kt
package org.findmyfam.models

import android.util.Base64
import org.json.JSONObject

/**
 * Self-contained invite token for joining a Marmot group.
 *
 * An invite encodes the relay URL, inviter's npub, and MLS group ID into
 * a compact base64-URL string that can be shared via messaging, QR code, etc.
 */
data class InviteCode(
    val relay: String,
    val inviterNpub: String,
    val groupId: String
) {
    /**
     * Encode the invite as a URL-safe base64 string.
     */
    fun encode(): String {
        val json = JSONObject().apply {
            put("relay", relay)
            put("inviterNpub", inviterNpub)
            put("groupId", groupId)
        }.toString()
        return Base64.encodeToString(json.toByteArray(Charsets.UTF_8), Base64.NO_WRAP)
    }

    /**
     * Wrap the invite in a famstr://invite/<code> deep-link URI.
     */
    fun asUri(): String {
        return "famstr://invite/${encode()}"
    }

    companion object {
        /**
         * Decode an invite from a base64-encoded string.
         */
        fun decode(encoded: String): InviteCode {
            val jsonBytes = Base64.decode(encoded, Base64.NO_WRAP)
            val obj = JSONObject(String(jsonBytes, Charsets.UTF_8))
            return InviteCode(
                relay = obj.getString("relay"),
                inviterNpub = obj.getString("inviterNpub"),
                groupId = obj.getString("groupId")
            )
        }

        /**
         * Extract an invite from a famstr://invite/<code> URI.
         * Also accepts a raw base64 string for backwards compatibility.
         */
        fun fromUri(uri: String): InviteCode {
            val prefix = "famstr://invite/"
            val code = if (uri.startsWith(prefix)) {
                uri.removePrefix(prefix)
            } else {
                uri
            }
            return decode(code)
        }

        /**
         * Build a famstr://addmember/<pubkeyHex>/<groupId> URI that the
         * invitee shares back with the inviter to request group admission.
         */
        fun approvalUri(pubkeyHex: String, groupId: String): String {
            return "famstr://addmember/$pubkeyHex/$groupId"
        }
    }
}

File: FindMyFamCore/Sources/FindMyFamCore/Models/InviteCode.swift
import Foundation

/// Self-contained invite token for joining a Marmot group.
///
/// An invite encodes the relay URL, inviter's npub, and MLS group ID into
/// a compact base64-URL string that can be shared via iMessage, QR code, etc.
public struct InviteCode: Codable, Equatable {

    /// Relay URL the invitee should connect to (e.g. "wss://relay.damus.io").
    public let relay: String

    /// Bech32 npub of the person who created the invite.
    public let inviterNpub: String

    /// MLS group identifier the invite is for.
    public let groupId: String

    public init(relay: String, inviterNpub: String, groupId: String) {
        self.relay = relay
        self.inviterNpub = inviterNpub
        self.groupId = groupId
    }

    // MARK: - Encoding

    /// Encode the invite as a URL-safe base64 string.
    public func encode() -> String {
        let data = try! JSONEncoder().encode(self)
        return data.base64EncodedString()
    }

    /// Wrap the invite in a `famstr://invite/<code>` deep-link URL.
    public func asURL() -> URL {
        URL(string: "famstr://invite/\(encode())")!
    }

    /// Decode an invite from a base64-encoded string.
    public static func decode(from encoded: String) throws -> InviteCode {
        guard let data = Data(base64Encoded: encoded) else {
            throw InviteError.invalidBase64
        }
        return try JSONDecoder().decode(InviteCode.self, from: data)
    }

    /// Extract an invite from a `famstr://invite/<code>` URL.
    /// Also accepts a raw base64 string for backwards compatibility.
    public static func from(url: URL) throws -> InviteCode {
        if url.scheme == "famstr", url.host == "invite",
           let code = url.pathComponents.dropFirst().first {
            return try decode(from: code)
        }
        return try decode(from: url.absoluteString)
    }

    // MARK: - Approval URL

    /// Build a `famstr://addmember/<pubkeyHex>/<groupId>` URL that the
    /// invitee shares back with the inviter to request group admission.
    public static func approvalURL(pubkeyHex: String, groupId: String) -> URL? {
        // groupId may contain characters that are invalid in a URL path component
        guard let encodedGroup = groupId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) else {
            return nil
        }
        return URL(string: "famstr://addmember/\(pubkeyHex)/\(encodedGroup)")
    }

    // MARK: - Errors

    public enum InviteError: LocalizedError {
        case invalidBase64

        public var errorDescription: String? {
            switch self {
            case .invalidBase64: return "Invalid invite code: not valid base64"
            }
        }
    }
}
